casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -94,25 +81,43 @@ object ProtoUtil {
     * requires a list to be returned. When we reach the goalFunc,
     * we return an empty list.
     */
-  def getCreatorJustificationAsListUntilGoalInMemory[F[_]: Monad](
+  def getCreatorJustificationAsListUntilGoalInMemory[F[_]: Sync](
       blockDag: BlockDagRepresentation[F],
       blockHash: BlockHash,
       goalFunc: BlockHash => Boolean = _ => false
-  ): F[List[BlockHash]] =
-    (for {
-      block <- OptionT(blockDag.lookup(blockHash))
-      creatorJustificationHash <- OptionT.fromOption(
-                                   block.justifications
-                                     .find(_.validator == block.sender)
-                                     .map(_.latestBlockHash)
-                                 )
-      creatorJustification <- OptionT(blockDag.lookup(creatorJustificationHash))
-      creatorJustificationAsList = if (goalFunc(creatorJustification.blockHash)) {
-        List.empty[BlockHash]
-      } else {
-        List(creatorJustification.blockHash)
-      }
-    } yield creatorJustificationAsList).fold(List.empty[BlockHash])(identity)
+  ): F[List[BlockHash]] = {
+    import coop.rchain.blockstorage.syntax._
+    val getJ = (h: BlockHash) =>
+      blockDag
+        .lookupUnsafe(h)
+        .map(_.justifications.map { case Justification(m, v) => Casper.Justification(m, v) })
+        .map(_.toSet)
+    val toJ = (h: BlockHash) =>
+      blockDag
+        .lookupUnsafe(h)
+        .map(m => Casper.Justification(m.blockHash, m.sender))
+
+    for {
+      r <- getCreatorJustification[F, BlockHash, Validator](blockHash, goalFunc, toJ, getJ)
+            .map(_.map(List(_)).getOrElse(List.empty))
+    } yield r
+
+  }
+
+  def getCreatorJustification[F[_]: Sync, A, V](
```

Comment:
Because `Justification` class is part of clique oracle, this function belongs there.